src/main/java/com/fruits/ecommerce/Services/Implementations/AuthImplementation.java

```diff
@@ -0,0 +1,393 @@
+        try {
+            emailService.sendAccountUnlockedEmail(user.getFirstName(), user.getEmail());
+            log.info("Account unlocked email sent successfully to: {}", user.getEmail());
+        } catch (MessagingException e) {
+            log.error("Failed to send account unlocked email to: {}. Error: {}", user.getEmail(), e.getMessage());
+            // Optionally handle the exception
+            throw new RuntimeException(e.getMessage());
+        }
+    }
+
+
+    private User getUserByIdentifier(String identifier) {
+        return userRepository.findByUsernameOrEmail(identifier, identifier)
+                .orElseThrow(() -> new UserNotFoundException("In-Valid UserAccount with: " + identifier));
+    }
+
+    // Private-Methods
+
+    /**
+     * Validate the new user's data.
+     *
+     * @param userDTO --->   The new user's information.
+     * @throws UsernameExistException   --->  If the username already exists.
+     * @throws EmailExistException      ---> If the email already exists.
+     * @throws InvalidUserDataException ---> If the user data is invalid.
+     */
+
+    // This method will validate the new user data before saving it
+    private void validateNewUser(UserDTO userDTO) {
+        if (!StringUtils.hasText(userDTO.getPassword())) {
+            throw new InvalidUserDataException("Password cannot be empty or null.");
+        }
+
+        if (userDTO.getFirstName() == null || userDTO.getLastName() == null ||
+                userDTO.getUsername() == null || userDTO.getEmail() == null) {
+            throw new InvalidUserDataException("All required fields must be provided.");
+        }
+
+        if (!isValidEmail(userDTO.getEmail())) {
+            throw new InvalidUserDataException("Invalid email format.");
+        }
+
+        if (userRepository.existsByUsername(userDTO.getUsername())) {
+            throw new UsernameExistException("Username already exists.");
+        }
+
+        if (userRepository.existsByEmail(userDTO.getEmail())) {
+            throw new EmailExistException("Email already exists.");
+        }
+        // Validate the user DTO using the validation annotations
+        Set<ConstraintViolation<UserDTO>> violations = validator.validate(userDTO);
+        if (!violations.isEmpty()) {
+            // Throw an InvalidUserDataException if there are any validation errors
+            StringBuilder errorMessage = new StringBuilder("Invalid user data: ");
+            for (ConstraintViolation<UserDTO> violation : violations) {
+                errorMessage.append(violation.getMessage()).append(", ");
+            }
+            throw new InvalidUserDataException(errorMessage.toString().trim());
+        }
+    }
+
+    // Helper method to validate the email address
+    private boolean isValidEmail(String email) {
+        // Regular expression to validate email format
+        String emailRegex = "^[a-zA-Z0-9_+&*-]+(?:\\.[a-zA-Z0-9_+&*-]+)*@(?:[a-zA-Z0-9-]+\\.)+[a-zA-Z]{2,7}$";
```

Comment:
Thanks for your valuable suggestion. I thought about it a lot and decided to take a slightly different approach, but one that achieves the same goal in a more efficient way. Instead of moving the regular expression to environment variables, I implemented a solution using the Apache Commons Validator library.